sdk/iOS/src/MSSyncContext.m

```diff
@@ -404,54 +406,53 @@ - (void) pullWithQuery:(MSQuery *)query queryId:(NSString *)queryId completion:(
 ///    If push fails, return error, else repeat while we have pending operations
 ///  Read from server using an MSQueuePullOperation
-- (void) pullWithQueryInternal:(MSQuery *)query queryId:(NSString *)queryId maxRecords:(NSInteger)maxRecords completion:(MSSyncBlock)completion
+- (MSQueuePullOperation *) pullWithQueryInternal:(MSQuery *)query queryId:(NSString *)queryId maxRecords:(NSInteger)maxRecords completion:(MSSyncBlock)completion
 {
-    dispatch_async(writeOperationQueue, ^{
-        // Before we can pull from the remote, we need to make sure out table doesn't having pending operations
-        NSArray *tableOps = [self.operationQueue getOperationsForTable:query.table.name item:nil];
-        if (tableOps.count > 0) {
-            [self pushWithCompletion:^(NSError *error) {
-                // For now we just abort the pull if the push failed to complete successfully
-                // Long term we can be smarter and check if our table succeeded
-                if (error) {
-                    if (completion) {
-                        [self.callbackQueue addOperationWithBlock:^{
-                            completion(error);
-                        }];
-                    }
-                } else {
-                    // Check again if we have new pending ops while we synced, and repeat as needed
-                    [self pullWithQueryInternal:query queryId:queryId maxRecords:maxRecords completion:completion];
-                }
-            }];
-            return;
-        } else {
-            // TODO: Allow users to cancel operations
-            MSQueuePullOperation *pull = [[MSQueuePullOperation alloc] initWithSyncContext:self
-                                                                                     query:query
-                                                                                   queryId:queryId
-                                                                                maxRecords:maxRecords
-                                                                             dispatchQueue:writeOperationQueue
-                                                                             callbackQueue:self.callbackQueue
-                                                                                completion:completion];
-            
-            
-            [pushQueue_ addOperation:pull];
-        }
-    });
+	MSQueuePullOperation *pull = [[MSQueuePullOperation alloc] initWithSyncContext:self
+																			 query:query
+																		   queryId:queryId
+																		maxRecords:maxRecords
+																	 dispatchQueue:writeOperationQueue
+																	 callbackQueue:self.callbackQueue
+																		completion:completion];
+	
+	// Before we can pull from the remote, we need to make sure out table doesn't having pending operations
+	NSArray *tableOps = [self.operationQueue getOperationsForTable:query.table.name item:nil];
```

Comment:
Originally I didn't see a way to return the pull operation from an async queue. Then I moved the pull constructor up to the first expression and I see I can add the async call back in. Will do it.